File: hu/codes/dart/chapter_backtracking/preorder_traversal_i_compact.dart
/**
 * File: preorder_traversal_i_compact.dart
 * Created Time: 2023-08-10
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

import '../utils/print_util.dart';
import '../utils/tree_node.dart';

/* Előrendű bejárás: 1. példa */
void preOrder(TreeNode? root, List<TreeNode> res) {
  if (root == null) {
    return;
  }
  if (root.val == 7) {
    // Megoldás rögzítése
    res.add(root);
  }
  preOrder(root.left, res);
  preOrder(root.right, res);
}

/* Főprogram */
void main() {
  TreeNode? root = listToTree([1, 7, 3, 4, 5, 6, 7]);
  print("\nInitialize binary tree");
  printTree(root);

  // Előrendű bejárás
  List<TreeNode> res = [];
  preOrder(root, res);

  print("\nOutput all nodes with value 7");
  print(List.generate(res.length, (i) => res[i].val));
}

File: hu/codes/cpp/chapter_backtracking/preorder_traversal_i_compact.cpp
/**
 * File: preorder_traversal_i_compact.cpp
 * Created Time: 2023-04-16
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

vector<TreeNode *> res;

/* Preorder bejárás: 1. példa */
void preOrder(TreeNode *root) {
    if (root == nullptr) {
        return;
    }
    if (root->val == 7) {
        // Megoldás rögzítése
        res.push_back(root);
    }
    preOrder(root->left);
    preOrder(root->right);
}

/* Főprogram */
int main() {
    TreeNode *root = vectorToTree(vector<int>{1, 7, 3, 4, 5, 6, 7});
    cout << "\nInitialize binary tree" << endl;
    printTree(root);

    // Preorder bejárás
    preOrder(root);

    cout << "\nOutput all nodes with value 7" << endl;
    vector<int> vals;
    for (TreeNode *node : res) {
        vals.push_back(node->val);
    }
    printVector(vals);
}
